### ISSUE TITLE
channel version of Racey test

### ISSUE THREAD

**Issue by devietti** [gatoWololo/rr_channel#36]
We should create a version of the Racey record-and-replay stress test [1] for Rust channels. Racey was written for shared memory, but I think we could port it as follows. Note: more elegant renditions of Racey into channel form are certainly welcome!

For each element of the `m` array in Racey (which is the key piece of shared state), we have two multi-producer/multi-consumer channels (so this is only for crossbeam, and maybe ipc?) and a dedicated thread, which collectively act like a "shared memory location". The pseudocode for one of these memory locations looks like this:

```
let mut u64 state = 0;
loop {
  let new_value = write_channel.try_recv();
  if new_value.is_there() { state = new_value; }
  if read_channel.is_empty() { read_channel.send(state); }
}
```
Basically the thread acts like a memory location, seeing if a write to its location comes in on the `write_channel`, and constantly sending out the current value on the `read_channel` since the number of reads-vs-writes is not guaranteed to be equal. Read and write channels should be bounded to avoid excessive space consumption. Maybe it would be faster to always keep the `read_channel` full, instead of non-empty.

Since we have a thread for each "memory location", we might need to shrink the size of `m` to 8 or 16 to avoid blowing up the system.

[1] http://pages.cs.wisc.edu/~markhill/racey.html

**Comment by devietti** [comment]
We need to support multiple consumers of a channel for this to work, and our rr_channel wrapper is still mpsc (even for mpmc crossbeam channels). Since Servo doesn't need multi-consumer support, it doesn't seem worth adding multi-consumer support just for this test. So this is low-priority until we encounter an application that needs multi-consumer support.
